gpytorch/models/deep_gps/deep_gp.py

```diff
@@ -0,0 +1,173 @@
+        self.output_dims = output_dims
+        self.num_samples = num_samples
+        self.last_layer = last_layer
+        self.register_added_loss_term("hidden_kl_divergence")
+
+    def forward(self, x):
+        raise NotImplementedError
+
+    def _reshape_input(self, inputs):
+        inputs = inputs.contiguous()
+        # Forward samples through the VariationalStrategy and return *samples* from q(f)
+        if inputs.dim() == 2:
+            # Assume new input entirely
+            inputs = inputs.unsqueeze(0)
+            inputs = inputs.expand(self.output_dims, inputs.size(-2), -1)
+        elif inputs.dim() == 3:
+            # Assume batch dim is samples, not output_dim
+            inputs = inputs.unsqueeze(0)
+            inputs = inputs.expand(self.output_dims, inputs.size(1), inputs.size(-2), -1)
+
+        return inputs
+
+    def __call__(self, inputs):
+        """
+        Forward data through this hidden GP layer.
+
+        If the input is 2 dimensional, we pass the input through each hidden GP, resulting in a `h x n` batch
+        Gaussian distribution. We then draw `s` samples from these Gaussians and reshape the result to be
+        `s x n x h` (e.g., h becomes the dimensionality of the output).
+
+        If the input is 3 dimensional, we assume that the input is `s x n x d`, e.g. the batch dimension
+        corresponds to the number of samples. We use this as the number of samples to draw, and just propagate
+        each sample through the hidden layer. The output will be `s x n x h`.
+
+        If the input is 4 dimensional, we assume that for some reason the user has already reshaped things to be
+        `h x s x n x d`. We reshape this internally to be `h x sn x d`, and the output will be `s x n x h`.
+
+        The goal of these last two points is that if you have a tensor `x` that is `n x d`, then:
+            >>> hidden_gp2(hidden_gp(x))
+
+        will just work, and return a tensor of size `s x n x h2`, where `h2` is the output dimensionality of
+        hidden_gp2. In this way, hidden GP layers are easily composable.
+        """
+        if inputs.size(-1) != self.input_dims:
+            raise RuntimeError(
+                f"Input shape did not match self.input_dims. Got total feature dims [{inputs.size(-1)}],"
+                f" expected [{self.input_dims}]"
+            )
+
+        if self.last_layer:
+            # Forward samples through the VariationalStrategy and return q(f)
+            last_layer_inputs = inputs.contiguous().view(-1, self.input_dims)
+            last_layer_inputs = last_layer_inputs.unsqueeze(0)
+            last_layer_inputs = last_layer_inputs.expand(
+                self.output_dims,
+                self.num_samples * inputs.size(-2),
+                self.input_dims
+            )
+
+            return AbstractVariationalGP.__call__(self, last_layer_inputs)
+        else:
+            inputs = self._reshape_input(inputs)
+            if inputs.dim() == 4:
```

Comment:
This can definitely be simplified now :)
Would it make sense for the samples dimension to be the outer dimension?